### PR TITLE
Refactor HoloLensVideoCapture to be more sane

### PR DESCRIPTION
**Pull request by jsundra**
Switching to use more managed state, and not requiring as explicit a warm/abort process as before.

### CODE REVIEW

File: Assets/Source/Application/States/Play/Hmd/HoloLensVideoCapture.cs [hunk 1]
```diff
@@ -70,213 +118,228 @@ public bool IsRecording
...
-                            Abort();
+                            Log.Info(this, "Entered VideoMode.");
+                            _currentState = CaptureState.InVideoMode;
+                            rtnToken.Succeed(Void.Instance);
                         }
                         else
                         {
-                            Log.Info(this, "Entered VideoMode.");
-                            _bootstrapper.BootstrapCoroutine(SucceedToken(_warmToken, Void.Instance));
+                            _currentState = CaptureState.Error;
```
Comment (thegoldenmule) [inline Assets/Source/Application/States/Play/Hmd/HoloLensVideoCapture.cs]: Let's log this case too.

File: Assets/Source/Application/States/Play/Hmd/HoloLensVideoCapture.cs [hunk 2]
```diff
@@ -70,213 +118,228 @@ public bool IsRecording
...
-                    _bootstrapper.BootstrapCoroutine(FailToken(rtnToken, exception));
-                });
+                    _currentState = CaptureState.Recording;
+                    rtnToken.Succeed(Void.Instance);
+                }
+                else
+                {
+                    _currentState = CaptureState.Error;
+                    rtnToken.Fail(new Exception(string.Format("Failure starting recording ({0})", result.hResult)));
+                }
```
Comment (thegoldenmule) [inline Assets/Source/Application/States/Play/Hmd/HoloLensVideoCapture.cs]: How about logs for both cases.

File: Assets/Source/Application/States/Play/Hmd/HoloLensVideoCapture.cs [hunk 3]
```diff
@@ -70,213 +118,228 @@ public bool IsRecording
...
-        /// <returns></returns>
-        private IEnumerator FailToken<T>(AsyncToken<T> token, Exception exception)
-        {
-            yield return null;
-            token.Fail(exception);
-        }
+            var rtnToken = new AsyncToken<Void>();
+            
+            Log.Info(this, "Refreshing VideoMode.");
+            _videoCapture.StopVideoModeAsync(stopResult =>
```
Comment (thegoldenmule) [inline Assets/Source/Application/States/Play/Hmd/HoloLensVideoCapture.cs]: No null check here scares me.

File: Assets/Source/Application/States/Play/Hmd/HoloLensVideoCapture.cs [hunk 4]
```diff
@@ -70,213 +118,228 @@ public bool IsRecording
...
+                    _videoCapture.StartVideoModeAsync(_cameraParameters, VideoCapture.AudioState.ApplicationAndMicAudio, startResult =>
+                    {
+                        if (startResult.success)
+                        {
+                            Log.Info(this, "VideoMode Refreshed.");
+                            rtnToken.Succeed(Void.Instance);
+                        }
+                        else
+                        {
+                            rtnToken.Fail(new Exception(string.Format("Failure entering VideoMode ({0})",
```
Comment (thegoldenmule) [inline Assets/Source/Application/States/Play/Hmd/HoloLensVideoCapture.cs]: log

**Comment by jsundra** [comment]
A lot of those errors would surface in the scripting APIs handling of them - but yeah it's better to have the capture system itself emitting the logs directly, and not sometime later in callbacks that may be out of order. Updated!

**Pull request merged**